Adding the 6th,7th and 8th precompiles and its tests, with the change…

madiazp opened the pull request:
…s of the

https://github.com/hyperledger/burrow/pull/1283 PR.


Signed-off-by: matias diaz <matias.dpz@gmail.com>

Go-ethereum use a [Cloudflare](https://github.com/cloudflare/bn256)  and a [deprecated google](https://go.googlesource.com/crypto/+/master/bn256/bn256.go)  package for [bn256 ](https://github.com/ethereum/go-ethereum/tree/master/crypto/bn256)modfied to work with libsnark. The cloudflare one have a BSD-3 license so is safe to use and modify.

With this package is possible to code[ the 6th,7th and 8th precompile contract](https://github.com/hyperledger/burrow/issues/1240).

I also notice the changes introduced in the #1283 PR. So I add some of them alongside with the precompile functions to make more easy to work with this PR (no credit stealing intended).

madiazp commented:
Also add some test for the 5,6,7 and 8th precompiles functions. For some reason the 5th one doesn't work with the changes of the #1283 PR. When the #1283 PR get merged I can add a Libsnark proof in the evm_test that will test the 5,6,7 and 8th at once with a real case use.

silasdavis commented on go.mod:
```diff
@@ -11,10 +11,12 @@ require (
 	github.com/asaskevich/govalidator v0.0.0-20180720115003-f9ffefc3facf // indirect
 	github.com/btcsuite/btcd v0.0.0-20190523000118-16327141da8c
 	github.com/cep21/xdgbasedir v0.0.0-20170329171747-21470bfc93b9
+	github.com/cloudflare/bn256 v0.0.0-20190523220833-828ba4f91854 // indirect
```
needs to be removed

silasdavis commented on go.mod:
```diff
@@ -11,10 +11,12 @@ require (
 	github.com/eapache/queue v1.1.0 // indirect
 	github.com/elgs/gojq v0.0.0-20160421194050-81fa9a608a13
 	github.com/elgs/gosplitargs v0.0.0-20161028071935-a491c5eeb3c8 // indirect
+	github.com/ethereum/go-ethereum v1.9.6
```
Also needs to be removed :)

silasdavis reviewed (changes requested):
Thanks for this Matias!

I think Clearmatics maintains this forked version of bn256 here: https://github.com/clearmatics/bn256. Let's use that as dependency rather than copying in the one from go-ethereum (though as you note there is no licensing issue with that, I'd just prefer to use a somewhat maintained fork).

If you rebase on master you'll get the other changes I have made, then I will give this another review and get it merged.

matiasADiazPerez committed:
Added the 6th,7th and 8th precompile, its tests and a zk-snark test for test them all together

Signed-off-by: matias diaz <matias.dpz@gmail.com>

madiazp commented:
Done ;)

Also fixed a problem with expMod. And added a proof snark test in evm_test.go

Your test of expMod on evm_test failed but the test of it in precompile_test passed and the snark_proof also passed so you may want to check it.

gregdhill commented:
@madiazp if you wouldn't mind re-basing this PR, I would be keen to merge this into master!

silasdavis commented:
@gregdhill perhaps you could pick up integrating this and maybe some of the other EVM updates we need.

madiazp commented:
Uh I don't know why the docker check failed. But again the test of expMod failed, you may want to check that personally. 
Sorry the delay
